staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go

```diff
@@ -17,333 +17,233 @@ limitations under the License.
-func (s *specAggregator) buildOpenAPISpec() (specToReturn *spec.Swagger, err error) {
-	specs := []openAPISpecInfo{}
-	for _, specInfo := range s.openAPISpecs {
-		if specInfo.spec == nil {
-			continue
-		}
-		// Copy the spec before removing the defaults.
-		localSpec := *specInfo.spec
-		localSpecInfo := *specInfo
-		localSpecInfo.spec = &localSpec
-		localSpecInfo.spec.Definitions = handler.PruneDefaults(specInfo.spec.Definitions)
-		specs = append(specs, localSpecInfo)
-	}
-	if len(specs) == 0 {
-		return &spec.Swagger{}, nil
-	}
-	sortByPriority(specs)
-	for _, specInfo := range specs {
-		if specToReturn == nil {
-			specToReturn = &spec.Swagger{}
-			*specToReturn = *specInfo.spec
-			// Paths and Definitions are set by MergeSpecsIgnorePathConflict
-			specToReturn.Paths = nil
-			specToReturn.Definitions = nil
-		}
-		if err := aggregator.MergeSpecsIgnorePathConflict(specToReturn, specInfo.spec); err != nil {
-			return nil, err
-		}
-	}
-
-	return specToReturn, nil
-}
-
-// updateOpenAPISpec aggregates all OpenAPI specs.  It is not thread-safe. The caller is responsible to hold proper locks.
-func (s *specAggregator) updateOpenAPISpec() error {
-	if s.openAPIVersionedService == nil {
-		return nil
+func (s *specAggregator) addLocalSpec(name string, spec cached.Data[*spec.Swagger]) {
+	service := v1.APIService{}
+	service.Name = name
+	info := &openAPISpecInfo{
+		apiService: service,
 	}
-	specToServe, err := s.buildOpenAPISpec()
-	if err != nil {
-		return err
-	}
-	return s.openAPIVersionedService.UpdateSpec(specToServe)
+	info.spec.Replace(spec)
+	s.openAPISpecs[name] = info
 }
 
-// tryUpdatingServiceSpecs tries updating openAPISpecs map with specified specInfo, and keeps the map intact
-// if the update fails.
-func (s *specAggregator) tryUpdatingServiceSpecs(specInfo *openAPISpecInfo) error {
-	if specInfo == nil {
-		return fmt.Errorf("invalid input: specInfo must be non-nil")
-	}
-	_, updated := s.openAPISpecs[specInfo.apiService.Name]
-	origSpecInfo, existedBefore := s.openAPISpecs[specInfo.apiService.Name]
-	s.openAPISpecs[specInfo.apiService.Name] = specInfo
-
-	// Skip aggregation if OpenAPI spec didn't change
-	if existedBefore && origSpecInfo != nil && origSpecInfo.etag == specInfo.etag {
-		return nil
-	}
-	klog.V(2).Infof("Updating OpenAPI spec because %s is updated", specInfo.apiService.Name)
-	defer func(start time.Time) {
-		duration := time.Since(start)
-		klog.V(2).Infof("Finished OpenAPI spec generation after %v", duration)
-
-		reason := "add"
-		if updated {
-			reason = "update"
+// buildMergeSpecLocked creates a new cached mergeSpec from the list of cached.
+func (s *specAggregator) buildMergeSpecLocked() cached.Data[*spec.Swagger] {
+	caches := make(map[string]cached.Data[*spec.Swagger], len(s.openAPISpecs))
+	for name, specInfo := range s.openAPISpecs {
+		caches[name] = cached.NewTransformer[*spec.Swagger](func(result cached.Result[*spec.Swagger]) cached.Result[*spec.Swagger] {
```

Comment:
This kind of transformation should not happen within `buildMergeSpecLocked()` imo.  This function recreates the Transformer, so an update to any CRD will cause the the `caches` map to be regenerated for all CRDs and PruneDefaults will be called for all CRDs, dropping any cached values that we may have had.

Reply:
Good catch, this is invalidating pruned specs that don't need to be pruned, thanks!!